core/src/main/java/io/questdb/cairo/map/FastMap.java

```diff
@@ -565,67 +502,255 @@ public void putDate(long value) {
-            checkSize(16);
             Unsafe.getUnsafe().putLong(appendAddress, lo);
             Unsafe.getUnsafe().putLong(appendAddress + Long.BYTES, hi);
             appendAddress += 16;
-            writeOffset();
         }
 
         @Override
         public void putLong256(Long256 value) {
-            checkSize(Long256.BYTES);
             Unsafe.getUnsafe().putLong(appendAddress, value.getLong0());
             Unsafe.getUnsafe().putLong(appendAddress + Long.BYTES, value.getLong1());
             Unsafe.getUnsafe().putLong(appendAddress + Long.BYTES * 2, value.getLong2());
             Unsafe.getUnsafe().putLong(appendAddress + Long.BYTES * 3, value.getLong3());
             appendAddress += Long256.BYTES;
-            writeOffset();
         }
 
         @Override
         public void putRecord(Record value) {
-            // noop
+            // no-op
         }
 
         @Override
         public void putShort(short value) {
-            checkSize(2);
             Unsafe.getUnsafe().putShort(appendAddress, value);
             appendAddress += 2;
-            writeOffset();
+        }
+
+        @Override
+        public void putTimestamp(long value) {
+            putLong(value);
+        }
+
+        @Override
+        public void skip(int bytes) {
+            appendAddress += bytes;
+        }
+
+        private MapValue createValue(FastMapValue value) {
+            commit();
+            // calculate hash remembering "key" structure
+            // [ len | value block | key offset block | key data block ]
+            int hashCode = hash();
+            int index = hashCode & mask;
+            long offset = getOffset(index);
+
+            if (offset < 0) {
+                return asNew(this, index, hashCode, value);
+            } else if (hashCode == getHashCode(index) && eq(offset)) {
+                return valueOf(kStart + offset, false, value);
+            } else {
+                return probe0(this, index, hashCode, value);
+            }
+        }
+
+        private MapValue findValue(FastMapValue value) {
+            commit();
+            int hashCode = hash();
+            int index = hashCode & mask;
+            long offset = getOffset(index);
+
+            if (offset < 0) {
+                return null;
+            } else if (eq(offset)) {
+                return valueOf(kStart + offset, false, value);
+            } else {
+                return probeReadOnly(this, index, hashCode, value);
+            }
+        }
+
+        protected void checkSize(int size) {
+            if (appendAddress + size > kLimit) {
+                resize(size);
+            }
+        }
+
+        protected void commit() {
+            // no-op
+        }
+
+        protected abstract boolean eq(long offset);
+
+        protected abstract int hash();
+    }
+
+    private class FixedSizeKey extends BaseKey {
+
+        public FixedSizeKey init() {
+            super.init();
+            checkSize(keySize + valueSize);
```

Comment:
The same applies to `FastMapValue` even before this patch. Nothing prevents you to write 4 `long` values instead of 4 `byte`s with chances of a segfault. So, `FastMap` as a whole is relatively unsafe. I can add `assert`s to check the size when running tests. WDYT?

Reply:
Added assertions in 52a88e678eefc07df8ccbd374e2fcd6778c7bbb5 and d90006e3532190773142fb08edfe2ed87f447b46